c/src/nnstreamer-capi-pipeline.c

```diff
@@ -1566,14 +1571,109 @@ ml_pipeline_src_input_data (ml_pipeline_src_h h, ml_tensors_data_h data,
   handle_exit (h);
 }
 
+/**
+ * @brief Internal function for appsrc callback - need_data.
+ */
+static void
+_pipe_src_cb_need_data (GstAppSrc * src, guint length, gpointer user_data)
+{
+  ml_pipeline_common_elem *src_h;
+  ml_pipeline_element *elem;
+  ml_pipeline_src_callbacks *src_cb = NULL;
+
+  src_h = (ml_pipeline_common_elem *) user_data;
+  elem = src_h->element;
+
+  g_mutex_lock (&elem->lock);
+  if (src_h->callback_info)
+    src_cb = &src_h->callback_info->src_cb;
+  g_mutex_unlock (&elem->lock);
+
+  if (src_cb && src_cb->need_data)
+    src_cb->need_data (src_h, length, src_h->callback_info->pdata);
```

Comment:
Move g_mutex_lock() from LINE 1950 to LINE 1594?
(callback function suppliers (API ```ml_pipeline_src_set_callback```'s caller) cannot access ```elem```'s lock anyway, right?)

Reply:
No, developer may call pipeline-src APIs (e.g. ml_pipeline_src_input_data()) in the callback with src_h.